File: 2023/d10.R
## This is a direct python-to-R translation of my python solution, so
## it does not feel fabulously R-ish, but now that I have my solution
## it's hard to see another way through it!
move <- list(north = c(-1, 0), south = c(1, 0), east = c(0, 1), west = c(0, -1))
turn <- list(north = list("|" = "north", "F" = "east",  "7" = "west"),
             south = list("|" = "south", "L" = "east",  "J" = "west"),
             east  = list("-" = "east",  "J" = "north", "7" = "south"),
             west  = list("-" = "west",  "L" = "north", "F" = "south"))
             
             
parse_input <- function(path) {
  map <- readLines(path)
  map <- matrix(unlist(strsplit(map, NULL)), length(map), byrow = TRUE)
  loop <- find_loop(c(which(map == "S", TRUE)), map)

  key <- paste(loop[nrow(loop), ] - loop[2, ] + 2, collapse = "")
  start <- c("24" = "-", "20" = "-", "42" = "|", "02" = "|",
             "13" = "F", "11" = "7", "31" = "J", "33" = "L")
  map[loop[1, , drop = FALSE]] <- start[[key]]

  map_clean <- array(NA_character_, dim(map))
  map_clean[loop] <- map[loop]

  list(loop = loop, map = map_clean)
}

find_loop <- function(start, map) {
  loop <- integer(2 * length(map)) # avoid concatenation in the loop
  for (direction in c("north", "south", "east", "west")) {
    tryCatch({
      pos <- start
      for (i in seq.int(1, 2 * length(map), 2)) {
        loop[c(i, i + 1L)] <- pos
        pos <- pos + move[[direction]]
        if (all(pos == start)) {
          return(matrix(loop[seq_len(i + 1)], ncol = 2, byrow = TRUE))
        }
        direction <- turn[[direction]][[map[pos[[1]], pos[[2]]]]]
      }
    }, error = function(e) NULL)
  }
}

part1 <- function(d) {
  nrow(d$loop) / 2
}

part2 <- function(d) {
  n_inside <- 0L
  map <- d$map
  for (i in seq_len(nrow(map))) {
    is_inside <- FALSE
    for (ch in map[i, ]) {
      if (is.na(ch)) {
        n_inside <- n_inside + is_inside
      } else if (ch == "|") {
        is_inside <- !is_inside
      } else if (ch == "F" || ch == "L") {
        start <- ch
      } else if ((ch == "7" && start == "L") || (ch == "J" && start == "F")) {
        is_inside = !is_inside
      }
    }
  }
  n_inside
}

stopifnot(part1(parse_input("test-input-d10-1.txt")) == 4)
stopifnot(part1(parse_input("test-input-d10-2.txt")) == 4)
stopifnot(part1(parse_input("test-input-d10-3.txt")) == 8)
stopifnot(part1(parse_input("test-input-d10-4.txt")) == 8)
stopifnot(part2(parse_input("test-input-d10-5.txt")) == 4)
stopifnot(part2(parse_input("test-input-d10-6.txt")) == 8)
stopifnot(part2(parse_input("test-input-d10-7.txt")) == 10)

d <- parse_input("input-d10.txt")
part1(d)
part2(d)

File: 2023/d10.py
from operator import add, sub

move = {
    "north": (-1,  0),
    "south": ( 1,  0),
    "east":  ( 0,  1),
    "west":  ( 0, -1)
}

turn = {
    ("north", "|"): "north",
    ("north", "F"): "east",
    ("north", "7"): "west",
    ("south", "|"): "south",
    ("south", "L"): "east",
    ("south", "J"): "west",
    ("east",  "-"): "east",
    ("east",  "J"): "north",
    ("east",  "7"): "south",
    ("west",  "-"): "west",
    ("west",  "L"): "north",
    ("west" , "F"): "south"
}


def parse_input(path):
    with open(path) as f:
        s = f.read()
    m = s.strip().split("\n")
    i = s.index("S")
    n = len(m[0]) + 1
    loop = find_loop((i // n, i % n), m)
    return loop, clean_map(loop, m)


def find_loop(start, m):
    for direction in ["east", "west", "north", "south"]:
        try:
            pos, loop = start, []
            while True:
                loop.append(pos)
                pos = tuple(map(add, pos, move[direction]))
                if pos == start:
                    return loop
                direction = turn[(direction, m[pos[0]][pos[1]])]
        except KeyError: # fail to look up a turn, this is junk
            pass


def clean_map(loop, m):
    ret = [[None] * len(m[0]) for _ in m]
    for r, c in loop:
        ret[r][c] = m[r][c]
    start = { # we don't actually see all of these in practice!
        (0, 2): "-", (0, -2): "-", (2, 0): "|", (-2, 0): "|",
        (-1, 1): "F", (-1, -1): "7", (1, -1): "J", (1, 1): "L"
    }
    ret[loop[0][0]][loop[0][1]] = start[tuple(map(sub, loop[1], loop[-1]))]
    return ret


def part1(d):
    return int(len(d[0]) / 2)


def part2(d):
    n_inside = 0
    for row in d[1]:
        is_inside = False
        for ch in row:
            if ch is None:
                n_inside = n_inside + int(is_inside)
            elif ch == "|":
                is_inside = not is_inside
            elif ch == "F" or ch == "L":
                start = ch
            elif (ch == "7" and start == "L") or (ch == "J" and start == "F"):
                is_inside = not is_inside
    return n_inside


assert part1(parse_input("test-input-d10-1.txt")) == 4
assert part1(parse_input("test-input-d10-2.txt")) == 4
assert part1(parse_input("test-input-d10-3.txt")) == 8
assert part1(parse_input("test-input-d10-4.txt")) == 8
assert part2(parse_input("test-input-d10-5.txt")) == 4
assert part2(parse_input("test-input-d10-6.txt")) == 8
assert part2(parse_input("test-input-d10-7.txt")) == 10

d = parse_input("input-d10.txt")
print(part1(d))
print(part2(d))
